docs: add noIndex:true docs for experimental query planning mode (#5883)
apollographql/router: docs/source/configuration/experimental_query_planning_mode.mdx

---
title: Experimental Query Planning Mode 
subtitle: Switch between legacy and native query planning
noIndex: true
---

The router (GraphOS Router and Apollo Router Core) is in the early stages of
transitioning to a native query planner, replacing the existing legacy planner.

As part of the efforts to ensure correctness and stability of the new planner,
v1.53.0 of the router enables both planners to run in parallel in order to
compare them. After the comparison, the router discards the native planner's results and
uses only the legacy planner to execute requests.

The native planner uses a single thread in the cold path of the router. It has a
bounded queue of 10 queries. If the queue is full, the router simply does not run the
comparison to avoid excessive resource consumption.

You can disable the native query planner by configuring your `router.yaml` to use just
`legacy` planning. You may want to disable it to avoid spikes in CPU utilization, for
example if an erroneous operation fails to complete planning in the native planner's
background thread.

> ```yaml title="router.yaml

Author: ```suggestion
```yaml title="router.yaml"
```

---
title: Experimental Query Planning Mode 
subtitle: Switch between legacy and native query planning
noIndex: true
---

> The router (GraphOS Router and Apollo Router Core) is in the early stages of transitioning to a native query planner,
> replacing the existing legacy planner.

Author: ```suggestion
The router (GraphOS Router and Apollo Router Core) is in the early stages of
transitioning to a native query planner, replacing the existing legacy planner.
```

As part of the efforts to ensure correctness and stability of the new planner,

> v1.53.0 of the router enables both planners to run in parallel in order

Reviewer: ```suggestion
v1.53.0 of the router enables both planners to run in parallel in order to
```

compare them. The native planner's result is discarded after the comparison and
only the legacy planner is used to execute requests.

The native planner uses a single thread in the cold path of the router and has a
bounded queue of 10. If the queue is full, the router simply does not run the

> comparison. This is done to avoid excessive resource consumption.

Author: Yes, the native query planner has its own bounded queue ✔️

> You may want to turn off this mode if you see a significant spike in CPU
> utilization. This can happen, if, for example, an erroneous operation fails to
> complete planning in the native planner's background thread, increasing this

Reviewer 2: ```suggestion
You can disable the native query planner by configuring your `router.yaml` to use just `legacy` planning. You may want to disable it to avoid spikes in CPU utilization, for example if an erroneous operation fails to complete planning in the native planner's background thread.
```

> thread's CPU utilization. You can swithc to just `legacy` planning in router.yaml:

Reviewer: ```suggestion
thread's CPU utilization. You can switch to just `legacy` planning in router.yaml:
```

Reviewer 2: ```suggestion
```

```yaml title="router.yaml
experimental_query_planner_mode: legacy
```

> The other variants for this feature flag are:

Reviewer 2: ```suggestion
The supported modes of `experimental_query_planner_mode` are the following:
```

* `new`. Enables only the native query planner.
* `both_best_effort` - default. Enables comparison between legacy and new native
   query planners. The legacy query planner is used for execution. If any
   unsupported features are detected, the router falls back to legacy with an
   `info` log.
* `legacy`. Enables only the legacy query planner.

> * `both`. Enables comparison between legacy and new native query planners.

Reviewer: The difference between `both` and `both_best_effort` is not clear to me? Why even have the `both` option around?

Reviewer 2: Is it that `both` won't fall back to legacy-only if unsupported features are detected?

Author: Yes, the `both` won't fall back. We only just internally introduced `both_best_effort` in the last week specifically for us to be able to run this canary testing for all deployments. I can remove `both_best_effort` (or I guess `both`) from this list if you think it's too confusing.

Author: ok sounds good!

> * `both`. Enables comparison between legacy and new native query planners.
>    query planners. The legacy query planner is used for execution. If any
>    unsupported features are detected, the router emits an error and reverts to
>    previous configuration. 

Author: ```suggestion
```
